text/0419-cdk-launchpads.md

```diff
@@ -0,0 +1,629 @@
+
+It can be done both using JSON/YAML:
+
+```json
+{
+  "launchpad": {
+    "bootstrap": {
+      "patch": {
+        "description": "Turn off image scanning on push",
+        "operations": [
+          {
+            "op": "replace",
+            "path": "/Resources/ContainerAssetsRepository/Properties/ImageScanningConfiguration/ScanOnPush",
+            "value": false
+          }
+        ]
+      }
+    }
+  }
+}
+```
+
+And CDK code:
+
+```ts
+new lpads.StandardCdkBootstrap(launchpad, 'CdkBootstrap', {
+  imageRepository: lpdas.CdkBootstrapImageRepository.default({
+    scanOnPush: false, // default: true
+  }),
+});
+```
+
+The `cdk-lpads` command contains a validator that makes sure any changes you make to the bootstrap template are correct,
+and satisfy the minimal requirements needed by CDK in order to still deploy using those bootstrap resources.
+
+#### Adding new bootstrap resources
+
+You can also add new resources to the Launchpad that you want to be shared between all applications deployed into this environment.
+Application developers can then easily import these resources into their CDK applications.
+
+You can add the new resources using JSON/YAML files:
+
+```json
+{
+  "launchpad": {
+    "resources": {
+      "cfn": {
+        "Resources": {
+          "PermissionBoundaryPolicy": {
+            "Type": "AWS::IAM::ManagedPolicy",
+            "Properties": {
+              "PolicyDocument": {
+                "Statement": [
+                  {
+                    "Action": "iam",
+                    "Effect": "Deny",
+                    "Resource": "*"
+                  }
+                ],
+                "Version": "2012-10-17"
+              },
+              "Description": "The managed policy used as the permission boundary"
+            }
+          }
+        }
+      },
+      "export": {
+        "permissionBoundaryPolicy": {
+          "Ref": "PermissionBoundaryPolicy"
+        }
+      }
+    }
+  }
+}
+```
+
+And also with CDK, using the same CDK libraries that are used for applications:
+
+```ts
+import * as iam from 'aws-cdk-lib/aws-iam';
+
+const managedPolicy = new iam.ManagedPolicy(launchpad, 'Boundary', {
+  statements: [
+    new iam.PolicyStatement({
+      effect: iam.Effect.DENY,
+      actions: ['iam:'],
+      resources: [''],
+    }),
+  ],
+});
+// surface this resource to application developers using this Launchpad
+// (seee below)
+launchpad.export(managedPolicy);
+```
+
+#### IAM Permission Boundaries
+
+Launchpads allow setting a given Managed Policy as the Permissions Boundary to be used by every application deployed into a given environment.
+
+You can do it through JSON/YAML:
+
+```json
+{
+  "launchpad": {
+    "iam": {
+      "permissionBoundary": {
+        "Ref": "PermissionBoundaryPolicy"
+      }
+    }
+  }
+}
+```
+
+Or in CDK code:
+
+```ts
+launchpad.setPermissionBoundaryPolicy(managedPolicy);
+```
+
+### Importing the Launchpad in CDK code
```

Comment:
I think the problem is that we need additional information to correctly do this importing. That's why, in the RFC, we have the `"export"` element in JSON, or the `export()` element in the CDK code of the Launchpad. Without this extra information, I don't think we'll have enough data to correctly generate CDK code from these resources.

Reply:
Is there physical information/known properties that launchpad information can be encoded too and looked up from like how bootstrap works today? We do this internally and using when using [CDK's ECS Service Extensions](https://www.npmjs.com/package/@aws-cdk-containers/ecs-service-extensions/v/2.0.0-alpha.17) (which has the notion of a shared "Environment").

Reply:
I think the problem is that you can't really cache these lookups, because the Launchpads can change with time. And so, you would need to re-do the lookup then, and we currently assume lookups can be cached forever (there's no easy way to invalidate a lookup, other than fiddling with the context yourself).

Code generation make this "invalidation" step easy (just re-run the generation part!).

Reply:
That's fair (and one of my issues with lookups in general). I guess my wish is just that it be more integrated than another step in the pipeline (and hopefully a cheap one), but maybe that can be a post-MVP thing?